// @sap/xsjs/lib/utils/date-utils.js
'use strict';

const NUMBER_OF_DIGITS_IN_MILLIS = 3;

module.exports = {
  dateToJsDate: dateToJsDate,
  timeToJsDate: timeToJsDate,
  dateTimeToJsDate: dateTimeToJsDate
};

function dateToJsDate(date, treatAsUTC) {
  return createDateObject({
    year: date.substring(0, 4),
    month: parseInt(date.substring(5, 7)) - 1,
    day: date.substring(8),
    hour: 0,
    minute: 0,
    second: 0,
    millis: 0
  }, treatAsUTC);
}

function timeToJsDate(time, treatAsUTC) {
  return createDateObject({
    year: 0,
    month: 0,
    day: 0,
    hour: time.substring(0, 2),
    minute: time.substring(3, 5),
    second: time.substring(6),
    millis: 0
  }, treatAsUTC);
}

function dateTimeToJsDate(dateTime, treatAsUTC) {
  var millis = dateTime.substring(20);
  millis += '0'.repeat(NUMBER_OF_DIGITS_IN_MILLIS - millis.length);

  return createDateObject({
    year: dateTime.substring(0, 4),
    month: parseInt(dateTime.substring(5, 7)) - 1,
    day: dateTime.substring(8, 10),
    hour: dateTime.substring(11, 13),
    minute: dateTime.substring(14, 16),
    second: dateTime.substring(17, 19),
    millis: millis
  }, treatAsUTC);
}

function createDateObject(values, treatAsUTC) {
  var date;
  if (treatAsUTC) {
    date = new Date(Date.UTC(values.year,
      values.month,
      values.day,
      values.hour,
      values.minute,
      values.second,
      values.millis));
  } else {
    date = new Date(values.year,
      values.month,
      values.day,
      values.hour,
      values.minute,
      values.second,
      values.millis);
  }

  // Workaround: two-digit years map to 1900-1999 which is not desired
  if (values.year >= 0 && values.year < 100) {
    if (treatAsUTC) {
      date.setUTCFullYear(values.year);
    } else {
      date.setFullYear(values.year);
    }
  }

  return date;
}
